File: design-patterns/go/memento/general/caretaker.go
// caretaker.go

package main

type Caretaker struct {
	mementoList []Memento
}

func NewCaretaker() (caretaker *Caretaker) {
	caretaker = &Caretaker{}
	return
}

func (caretaker *Caretaker) Add(memento Memento) {
	caretaker.mementoList = append(caretaker.mementoList, memento)
}

func (caretaker *Caretaker) GetByIndex(index int) Memento {
	return caretaker.mementoList[index]
}

func (caretaker *Caretaker) GetCurrent() Memento {
	return caretaker.mementoList[len(caretaker.mementoList)-1]
}

func (caretaker *Caretaker) Undo() {
	caretaker.mementoList = caretaker.mementoList[:len(caretaker.mementoList)-1]
}

File: design-patterns/php/memento/general/Caretaker.php
<?php
// Caretaker.php

namespace BigBoxCode\DesignPattern\Memento\General;


class Caretaker {

    private array $mementoList = [];

    public function add(Memento $memento): void {
        array_push($this->mementoList, $memento);
    }

    public function getByIndex(int $index): Memento {
        return $this->mementoList[$index];
    }

    public function getCurrent(): Memento {
        return $this->mementoList[count($this->mementoList) - 1];
    }

    public function undo(): void {
        unset($this->mementoList[count($this->mementoList) - 1]);
    }
}